api/update_test.go

```diff
@@ -0,0 +1,86 @@
+package api
+
+import (
+	"bytes"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/chromium/hstspreload"
+	"github.com/chromium/hstspreload/chromium/preloadlist"
+)
+
+const ()
+
+// TestPolicyType tests that PolicyType is populated within the database when the Update endpoint is called.
+func TestPolicyType(t *testing.T) {
+	api, mockController, mockHstspreload, mockPreloadlist := mockAPI(0 * time.Second)
+
+	api.bulkPreloaded["removal-preloaded-bulk-eligible.test"] = true
+	api.bulkPreloaded["removal-not-preloaded-bulk-eligible.test"] = true
+	api.bulkPreloaded["removal-preloaded-bulk-ineligible.test"] = true
+
+	TestPreloadableResponses := map[string]hstspreload.Issues{
+		"garron.net":                      emptyIssues,
+		"badssl.com":                      issuesWithWarnings,
+		"example.com":                     issuesWithErrors,
+		"removal-pending-eligible.test":   emptyIssues,
+		"removal-pending-ineligible.test": emptyIssues,
+	}
+	TestRemovableResponses := map[string]hstspreload.Issues{
+		"removal-preloaded-bulk-eligible.test":     emptyIssues,
+		"removal-preloaded-not-bulk-eligible.test": emptyIssues,
+		"removal-preloaded-bulk-ineligible.test":   issuesWithErrors,
+		"removal-pending-eligible.test":            emptyIssues,
+		"removal-pending-ineligible.test":          issuesWithErrors,
+	}
+
+	TestPreloadlist := preloadlist.PreloadList{Entries: []preloadlist.Entry{
+		{Name: "garron.net", Mode: preloadlist.ForceHTTPS, IncludeSubDomains: true, Policy: preloadlist.Bulk18Weeks},
+		{Name: "chromium.org", Mode: preloadlist.ForceHTTPS, IncludeSubDomains: false, Policy: preloadlist.Custom},
+		{Name: "removal-preloaded-bulk-eligible.test", Mode: preloadlist.ForceHTTPS, IncludeSubDomains: true, Policy: preloadlist.UnspecifiedPolicyType},
+		{Name: "removal-preloaded-not-bulk-eligible.test", Mode: preloadlist.ForceHTTPS, IncludeSubDomains: true, Policy: preloadlist.Bulk1Year},
+		{Name: "removal-preloaded-bulk-ineligible.test", Mode: preloadlist.ForceHTTPS, IncludeSubDomains: true, Policy: preloadlist.Test},
+		{Name: "godoc.og", Mode: "", IncludeSubDomains: true, Policy: preloadlist.Custom},
+		{Name: "dev", Mode: preloadlist.ForceHTTPS, IncludeSubDomains: true, Policy: preloadlist.UnspecifiedPolicyType},
+	}}
+
+	mockHstspreload.preloadableResponses = TestPreloadableResponses
+	mockHstspreload.removableResponses = TestRemovableResponses
+	mockPreloadlist.list = TestPreloadlist
+
+	mockController.FailCalls = (failNone & failDatabase) != 0
+	mockPreloadlist.failCalls = (failNone & failChromiumpreload) != 0
+
+	w := httptest.NewRecorder()
+	w.Body = &bytes.Buffer{}
+
+	r, err := http.NewRequest("GET", "", nil)
+	if err != nil {
+		t.Fatalf("[%s] %s", "NewRequest Failed", err)
+	}
+
+	api.Update(w, r)
+
+	states, err := api.database.AllDomainStates()
+	if err != nil {
+		t.Fatalf("Couldn't get the states of all domains in the database.")
+	}
+
+	expectedPolicies := map[string]preloadlist.PolicyType{
```

Comment:
It looks like all of these are entries that are on the preload list. AllDomainStates will also return domains that aren't on the preload list, so you should test somewhere (maybe in a database test) that it returns the correct PolicyType for domains that aren't on the list.

Also, afaict the call to `Update` inserts all of these domains into the datastore. It would be good to test what happens when a domain is already in the datastore (e.g. is StatusPending) and then gets updated to StatusPreloaded.

Reply:
Done